File: java/src/main/java/com/hibana/samples/Example10_ErrorHandling.java
package com.hibana.samples;

import com.openai.client.OpenAIClient;
import com.openai.client.okhttp.OpenAIOkHttpClient;
import com.openai.core.http.StreamResponse;
import com.openai.errors.*;
import com.openai.models.ChatCompletion;
import com.openai.models.ChatCompletionCreateParams;
import com.openai.models.ChatCompletionMessage;

/**
 * 10 - Error Handling
 *
 * This example demonstrates proper error handling when using the Hibana API.
 * It covers common errors, retry strategies, and best practices for building
 * robust applications.
 */
public class Example10_ErrorHandling {

    private static final String API_KEY = "YOUR_API_KEY";
    private static final String BASE_URL = "https://api-ai.hibanacloud.com/v1";

    public static void main(String[] args) {
        try {
            // handleAuthenticationError();  // Uncomment to test
            // handleModelNotFound();  // Uncomment to test
            // handleRateLimit();  // Uncomment to test
            handleInsufficientBalance();
            System.out.println("\n");
            retryWithExponentialBackoff();
            System.out.println("\n");
            comprehensiveErrorHandler();
            System.out.println("\n");
            validateBeforeRequest();
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static void handleAuthenticationError() {
        /**
         * Handle authentication errors (invalid API key)
         */

        System.out.println("=".repeat(60));
        System.out.println("Error Handling - Authentication");
        System.out.println("=".repeat(60));

        // Create client with invalid API key
        OpenAIClient badClient = OpenAIOkHttpClient.builder()
                .apiKey("INVALID_KEY_123")
                .baseUrl(BASE_URL)
                .build();

        try {
            ChatCompletionCreateParams params = ChatCompletionCreateParams.builder()
                    .model("gpt-5-nano")
                    .addMessage(ChatCompletionMessage.ofChatCompletionUserMessageParam(
                            ChatCompletionMessage.ChatCompletionUserMessageParam.builder()
                                    .content("Hello")
                                    .build()
                    ))
                    .build();

            badClient.chat().completions().create(params);

        } catch (AuthenticationError e) {
            System.out.println("\n✗ Authentication Error Caught!");
            System.out.println("Error message: " + e.getMessage());
            System.out.println("\nSolution:");
            System.out.println("  1. Check that your API key is correct");
            System.out.println("  2. Verify the key is active in your dashboard");
            System.out.println("  3. Ensure proper Bearer token format");
        }
    }

    private static void handleModelNotFound() {
        /**
         * Handle errors when requesting unavailable models
         */

        System.out.println("=".repeat(60));
        System.out.println("Error Handling - Model Not Found");
        System.out.println("=".repeat(60));

        OpenAIClient client = OpenAIOkHttpClient.builder()
                .apiKey(API_KEY)
                .baseUrl(BASE_URL)
                .build();

        try {
            ChatCompletionCreateParams params = ChatCompletionCreateParams.builder()
                    .model("non-existent-model-xyz")
                    .addMessage(ChatCompletionMessage.ofChatCompletionUserMessageParam(
                            ChatCompletionMessage.ChatCompletionUserMessageParam.builder()
                                    .content("Hello")
                                    .build()
                    ))
                    .build();

            client.chat().completions().create(params);

        } catch (NotFoundError | BadRequestError e) {
            System.out.println("\n✗ Model Not Found Error!");
            System.out.println("Error: " + e.getMessage());
            System.out.println("\nSolution:");
            System.out.println("  1. Use client.models().list() to see available models");
            System.out.println("  2. Check for typos in model name");
            System.out.println("  3. Verify model is enabled in your account");
        }
    }

    private static void handleRateLimit() {
        /**
         * Handle rate limiting errors
         */

        System.out.println("=".repeat(60));
        System.out.println("Error Handling - Rate Limiting");
        System.out.println("=".repeat(60));

        OpenAIClient client = OpenAIOkHttpClient.builder()
                .apiKey(API_KEY)
                .baseUrl(BASE_URL)
                .build();

        System.out.println("\nSimulating rate limit scenario...");

        try {
            for (int i = 0; i < 5; i++) {
                ChatCompletionCreateParams params = ChatCompletionCreateParams.builder()
                        .model("gpt-5-nano")
                        .addMessage(ChatCompletionMessage.ofChatCompletionUserMessageParam(
                                ChatCompletionMessage.ChatCompletionUserMessageParam.builder()
                                        .content("Request " + (i + 1))
                                        .build()
                        ))
                        .maxTokens(10000L)
                        .build();

                client.chat().completions().create(params);
                System.out.println("Request " + (i + 1) + ": Success");
                Thread.sleep(100);  // Small delay
            }

        } catch (RateLimitError e) {
            System.out.println("\n✗ Rate Limit Error: " + e.getMessage());
            System.out.println("\nSolution:");
            System.out.println("  1. Implement exponential backoff");
            System.out.println("  2. Add delays between requests");
            System.out.println("  3. Monitor rate limits for your tier");
            System.out.println("  4. Consider upgrading your plan");
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    private static void handleInsufficientBalance() {
        /**
         * Handle insufficient balance errors
         */

        System.out.println("=".repeat(60));
        System.out.println("Error Handling - Insufficient Balance");
        System.out.println("=".repeat(60));

        OpenAIClient client = OpenAIOkHttpClient.builder()
                .apiKey(API_KEY)
                .baseUrl(BASE_URL)
                .build();

        System.out.println("\nChecking for balance errors...");

        try {
            ChatCompletionCreateParams params = ChatCompletionCreateParams.builder()
                    .model("gpt-5-nano")
                    .addMessage(ChatCompletionMessage.ofChatCompletionUserMessageParam(
                            ChatCompletionMessage.ChatCompletionUserMessageParam.builder()
                                    .content("Test")
                                    .build()
                    ))
                    .maxTokens(10000L)
                    .build();

            client.chat().completions().create(params);
            System.out.println("✓ Request successful - sufficient balance");

        } catch (OpenAIException e) {
            if (e.getMessage().toLowerCase().contains("insufficient") ||
                    e.getMessage().toLowerCase().contains("balance")) {
                System.out.println("\n✗ Insufficient Balance: " + e.getMessage());
                System.out.println("\nSolution:");
                System.out.println("  1. Check balance: GET /v1/user/balance");
                System.out.println("  2. Recharge your account");
                System.out.println("  3. Use lower-cost models");
            }
        }
    }

    private static void retryWithExponentialBackoff() {
        /**
         * Implement retry logic with exponential backoff
         */

        System.out.println("=".repeat(60));
        System.out.println("Retry Strategy - Exponential Backoff");
        System.out.println("=".repeat(60));

        OpenAIClient client = OpenAIOkHttpClient.builder()
                .apiKey(API_KEY)
                .baseUrl(BASE_URL)
                .build();

        int maxRetries = 3;
        long baseDelay = 1000;  // milliseconds

        for (int attempt = 0; attempt < maxRetries; attempt++) {
            try {
                System.out.println("\nAttempt " + (attempt + 1) + "/" + maxRetries + "...");

                ChatCompletionCreateParams params = ChatCompletionCreateParams.builder()
                        .model("gpt-5-nano")
                        .addMessage(ChatCompletionMessage.ofChatCompletionUserMessageParam(
                                ChatCompletionMessage.ChatCompletionUserMessageParam.builder()
                                        .content("Hello")
                                        .build()
                        ))
                        .maxTokens(10000L)
                        .build();

                ChatCompletion response = client.chat().completions().create(params);

                System.out.println("✓ Success!");
                System.out.println("Response: " + response.choices().get(0).message().content().orElse(""));
                break;

            } catch (RateLimitError e) {
                if (attempt < maxRetries - 1) {
                    long delay = baseDelay * (long) Math.pow(2, attempt);
                    System.out.println("✗ Rate limited. Retrying in " + (delay / 1000) + " seconds...");
                    try {
                        Thread.sleep(delay);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                    }
                } else {
                    System.out.println("✗ Failed after " + maxRetries + " attempts");
                    throw e;
                }
            } catch (OpenAIException e) {
                if (attempt < maxRetries - 1) {
                    long delay = baseDelay * (long) Math.pow(2, attempt);
                    System.out.println("✗ API Error: " + e.getMessage());
                    System.out.println("Retrying in " + (delay / 1000) + " seconds...");
                    try {
                        Thread.sleep(delay);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                    }
                } else {
                    System.out.println("✗ Failed after " + maxRetries + " attempts");
                    throw e;
                }
            }
        }
    }

    private static void comprehensiveErrorHandler() {
        /**
         * A comprehensive error handling wrapper
         */

        System.out.println("=".repeat(60));
        System.out.println("Comprehensive Error Handler");
        System.out.println("=".repeat(60));

        System.out.println("\nTesting with valid request:");
        ChatCompletion result = safeApiCall("gpt-5-nano", "Say hello in 3 words");

        if (result != null) {
            System.out.println("✓ Success: " + result.choices().get(0).message().content().orElse(""));
        } else {
            System.out.println("Request failed gracefully");
        }
    }

    private static ChatCompletion safeApiCall(String model, String userMessage) {
        /**
         * Wrapper function with comprehensive error handling
         */

        OpenAIClient client = OpenAIOkHttpClient.builder()
                .apiKey(API_KEY)
                .baseUrl(BASE_URL)
                .build();

        try {
            ChatCompletionCreateParams params = ChatCompletionCreateParams.builder()
                    .model(model)
                    .addMessage(ChatCompletionMessage.ofChatCompletionUserMessageParam(
                            ChatCompletionMessage.ChatCompletionUserMessageParam.builder()
                                    .content(userMessage)
                                    .build()
                    ))
                    .maxTokens(10000L)
                    .build();

            return client.chat().completions().create(params);

        } catch (AuthenticationError e) {
            System.out.println("✗ Authentication Error: " + e.getMessage());
            System.out.println("Check your API key.");
            return null;
        } catch (NotFoundError e) {
            System.out.println("✗ Not Found: " + e.getMessage());
            System.out.println("Model may not exist or be unavailable.");
            return null;
        } catch (RateLimitError e) {
            System.out.println("✗ Rate Limit: " + e.getMessage());
            System.out.println("Too many requests. Please wait and try again.");
            return null;
        } catch (BadRequestError e) {
            System.out.println("✗ Bad Request: " + e.getMessage());
            System.out.println("Check your request parameters.");
            return null;
        } catch (InternalServerError e) {
            System.out.println("✗ API Error: " + e.getMessage());
            System.out.println("An error occurred on the server side.");
            return null;
        } catch (OpenAIException e) {
            System.out.println("✗ Unexpected Error: " + e.getMessage());
            return null;
        }
    }

    private static void validateBeforeRequest() {
        /**
         * Validate inputs before making API request
         */

        System.out.println("=".repeat(60));
        System.out.println("Input Validation");
        System.out.println("=".repeat(60));

        // Test validation
        System.out.println("\nTest 1: Invalid model");
        validateAndCall("invalid-model", "Hello");

        System.out.println("\nTest 2: Empty input");
        validateAndCall("gpt-5-nano", "");

        System.out.println("\nTest 3: Valid request");
        ChatCompletion result = validateAndCall("gpt-5-nano", "Say hi");
        if (result != null) {
            System.out.println("Response: " + result.choices().get(0).message().content().orElse(""));
        }
    }

    private static ChatCompletion validateAndCall(String model, String userInput) {
        /**
         * Validate inputs before API call
         */

        // Validate model name
        String[] validModels = {"gpt-5-nano", "claude-haiku-4-5", "deepseek-chat", "gemini-2.5-flash-lite"};
        boolean validModel = false;
        for (String valid : validModels) {
            if (valid.equals(model)) {
                validModel = true;
                break;
            }
        }

        if (!validModel) {
            System.out.println("✗ Invalid model: " + model);
            System.out.println("Valid models: " + String.join(", ", validModels));
            return null;
        }

        // Validate user input
        if (userInput == null || userInput.trim().isEmpty()) {
            System.out.println("✗ Empty user input");
            return null;
        }

        if (userInput.length() > 10000) {
            System.out.println("✗ Input too long (max 10,000 characters)");
            return null;
        }

        // All validations passed
        System.out.println("✓ Validation passed");

        OpenAIClient client = OpenAIOkHttpClient.builder()
                .apiKey(API_KEY)
                .baseUrl(BASE_URL)
                .build();

        try {
            ChatCompletionCreateParams params = ChatCompletionCreateParams.builder()
                    .model(model)
                    .addMessage(ChatCompletionMessage.ofChatCompletionUserMessageParam(
                            ChatCompletionMessage.ChatCompletionUserMessageParam.builder()
                                    .content(userInput)
                                    .build()
                    ))
                    .maxTokens(10000L)
                    .build();

            return client.chat().completions().create(params);
        } catch (OpenAIException e) {
            System.out.println("✗ Error: " + e.getMessage());
            return null;
        }
    }
}

File: csharp/Example10_ErrorHandling.cs
/*
 * 10 - Error Handling
 *
 * This example demonstrates proper error handling when using the Hibana API.
 * It covers common errors, retry strategies, and best practices for building
 * robust applications.
 */

using System;
using System.ClientModel;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using OpenAI;
using OpenAI.Chat;

namespace Hibana.Samples
{
    public class Example10_ErrorHandling
    {
        private static readonly string ApiKey = "YOUR_API_KEY";
        private static readonly string BaseUrl = "https://api-ai.hibanacloud.com/v1";

        public static async Task Main(string[] args)
        {
            try
            {
                // Example 1: Authentication errors
                // await HandleAuthenticationError();  // Uncomment to test

                // Example 2: Model not found
                // await HandleModelNotFound();  // Uncomment to test

                // Example 3: Rate limiting
                // await HandleRateLimit();  // Uncomment to test

                // Example 4: Insufficient balance
                await HandleInsufficientBalance();

                // Example 5: Retry with backoff
                await RetryWithExponentialBackoff();

                // Example 6: Timeout handling
                // await HandleTimeout();  // Uncomment to test

                // Example 7: Comprehensive handler
                await ComprehensiveErrorHandler();

                // Example 8: Input validation
                await ValidateBeforeRequest();

                Console.WriteLine("\n" + new string('=', 60));
                Console.WriteLine("Error handling examples completed!");
                Console.WriteLine(new string('=', 60));
            }
            catch (Exception ex)
            {
                Console.WriteLine($"\nUnexpected error in main: {ex.Message}");
            }
        }

        /// <summary>
        /// Handle authentication errors (invalid API key)
        /// </summary>
        private static async Task HandleAuthenticationError()
        {
            Console.WriteLine(new string('=', 60));
            Console.WriteLine("Error Handling - Authentication");
            Console.WriteLine(new string('=', 60));

            // Create client with invalid API key
            var badClient = new OpenAIClient(new ApiKeyCredential("INVALID_KEY_123"), new OpenAIClientOptions
            {
                Endpoint = new Uri(BaseUrl)
            });

            var chatClient = badClient.GetChatClient("gpt-5-nano");

            try
            {
                await chatClient.CompleteChatAsync(
                    messages: new[] { new UserChatMessage("Hello") }
                );
            }
            catch (Exception ex)
            {
                Console.WriteLine("\n✗ Authentication Error Caught!");
                Console.WriteLine($"Error message: {ex.Message}");
                Console.WriteLine("\nSolution:");
                Console.WriteLine("  1. Check that your API key is correct");
                Console.WriteLine("  2. Verify the key is active in your dashboard");
                Console.WriteLine("  3. Ensure proper Bearer token format");
            }
        }

        /// <summary>
        /// Handle errors when requesting unavailable models
        /// </summary>
        private static async Task HandleModelNotFound()
        {
            Console.WriteLine("\n" + new string('=', 60));
            Console.WriteLine("Error Handling - Model Not Found");
            Console.WriteLine(new string('=', 60));

            var client = new OpenAIClient(new ApiKeyCredential(ApiKey), new OpenAIClientOptions
            {
                Endpoint = new Uri(BaseUrl)
            });

            var chatClient = client.GetChatClient("non-existent-model-xyz");

            try
            {
                await chatClient.CompleteChatAsync(
                    messages: new[] { new UserChatMessage("Hello") }
                );
            }
            catch (Exception ex)
            {
                Console.WriteLine("\n✗ Model Not Found Error!");
                Console.WriteLine($"Error: {ex.Message}");
                Console.WriteLine("\nSolution:");
                Console.WriteLine("  1. Use modelClient.GetModelsAsync() to see available models");
                Console.WriteLine("  2. Check for typos in model name");
                Console.WriteLine("  3. Verify model is enabled in your account");
            }
        }

        /// <summary>
        /// Handle rate limiting errors
        /// </summary>
        private static async Task HandleRateLimit()
        {
            Console.WriteLine("\n" + new string('=', 60));
            Console.WriteLine("Error Handling - Rate Limiting");
            Console.WriteLine(new string('=', 60));

            Console.WriteLine("\nSimulating rate limit scenario...");

            var client = new OpenAIClient(new ApiKeyCredential(ApiKey), new OpenAIClientOptions
            {
                Endpoint = new Uri(BaseUrl)
            });

            var chatClient = client.GetChatClient("gpt-5-nano");

            try
            {
                // This might trigger rate limiting if called too frequently
                for (int i = 0; i < 5; i++)
                {
                    await chatClient.CompleteChatAsync(
                        messages: new[] { new UserChatMessage($"Request {i + 1}") },
                        options: new ChatCompletionOptions { MaxOutputTokenCount = 10000 }
                    );
                    Console.WriteLine($"Request {i + 1}: Success");
                    await Task.Delay(100);  // Small delay
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"\n⚠ Rate Limit Error: {ex.Message}");
                Console.WriteLine("\nSolution:");
                Console.WriteLine("  1. Implement exponential backoff");
                Console.WriteLine("  2. Add delays between requests");
                Console.WriteLine("  3. Monitor rate limits for your tier");
                Console.WriteLine("  4. Consider upgrading your plan");
            }
        }

        /// <summary>
        /// Handle insufficient balance errors
        /// </summary>
        private static async Task HandleInsufficientBalance()
        {
            Console.WriteLine("\n" + new string('=', 60));
            Console.WriteLine("Error Handling - Insufficient Balance");
            Console.WriteLine(new string('=', 60));

            Console.WriteLine("\nChecking for balance errors...");

            var client = new OpenAIClient(new ApiKeyCredential(ApiKey), new OpenAIClientOptions
            {
                Endpoint = new Uri(BaseUrl)
            });

            var chatClient = client.GetChatClient("gpt-5-nano");

            try
            {
                // This would fail if balance is insufficient
                await chatClient.CompleteChatAsync(
                    messages: new[] { new UserChatMessage("Test") },
                    options: new ChatCompletionOptions { MaxOutputTokenCount = 10000 }
                );
                Console.WriteLine("✓ Request successful - sufficient balance");
            }
            catch (Exception ex)
            {
                string errorMsg = ex.Message.ToLower();
                if (errorMsg.Contains("insufficient") || errorMsg.Contains("balance"))
                {
                    Console.WriteLine($"\n✗ Insufficient Balance: {ex.Message}");
                    Console.WriteLine("\nSolution:");
                    Console.WriteLine("  1. Check balance: GET /v1/user/balance");
                    Console.WriteLine("  2. Recharge your account");
                    Console.WriteLine("  3. Use lower-cost models");
                }
            }
        }

        /// <summary>
        /// Implement retry logic with exponential backoff
        /// </summary>
        private static async Task RetryWithExponentialBackoff()
        {
            Console.WriteLine("\n" + new string('=', 60));
            Console.WriteLine("Retry Strategy - Exponential Backoff");
            Console.WriteLine(new string('=', 60));

            var client = new OpenAIClient(new ApiKeyCredential(ApiKey), new OpenAIClientOptions
            {
                Endpoint = new Uri(BaseUrl)
            });

            var chatClient = client.GetChatClient("gpt-5-nano");

            int maxRetries = 3;
            int baseDelay = 1000;  // milliseconds

            for (int attempt = 0; attempt < maxRetries; attempt++)
            {
                try
                {
                    Console.WriteLine($"\nAttempt {attempt + 1}/{maxRetries}...");

                    var response = await chatClient.CompleteChatAsync(
                        messages: new[] { new UserChatMessage("Hello") },
                        options: new ChatCompletionOptions { MaxOutputTokenCount = 10000 }
                    );

                    Console.WriteLine("✓ Success!");
                    Console.WriteLine($"Response: {response.Value.Content[0].Text}");
                    break;
                }
                catch (Exception ex)
                {
                    if (attempt < maxRetries - 1)
                    {
                        int delay = baseDelay * (int)Math.Pow(2, attempt);  // Exponential backoff
                        Console.WriteLine($"⚠ Error. Retrying in {delay / 1000} seconds...");
                        await Task.Delay(delay);
                    }
                    else
                    {
                        Console.WriteLine($"✗ Failed after {maxRetries} attempts");
                        throw;
                    }
                }
            }
        }

        /// <summary>
        /// Handle request timeout errors
        /// </summary>
        private static async Task HandleTimeout()
        {
            Console.WriteLine("\n" + new string('=', 60));
            Console.WriteLine("Error Handling - Timeouts");
            Console.WriteLine(new string('=', 60));

            // Set a short timeout for demonstration
            var timeoutClient = new OpenAIClient(new ApiKeyCredential(ApiKey), new OpenAIClientOptions
            {
                Endpoint = new Uri(BaseUrl)
                // Note: .NET OpenAI client timeout configuration may vary
            });

            var chatClient = timeoutClient.GetChatClient("gpt-5-nano");

            try
            {
                var cts = new CancellationTokenSource(TimeSpan.FromMilliseconds(1)); // Very short timeout
                await chatClient.CompleteChatAsync(
                    messages: new[] { new UserChatMessage("Hello") },
                    cancellationToken: cts.Token
                );
            }
            catch (OperationCanceledException ex)
            {
                Console.WriteLine($"\n⚠ Timeout Error: {ex.Message}");
                Console.WriteLine("\nSolution:");
                Console.WriteLine("  1. Increase timeout value");
                Console.WriteLine("  2. Check network connection");
                Console.WriteLine("  3. Try again later if server is slow");
            }
        }

        /// <summary>
        /// A comprehensive error handling wrapper
        /// </summary>
        private static async Task ComprehensiveErrorHandler()
        {
            Console.WriteLine("\n" + new string('=', 60));
            Console.WriteLine("Comprehensive Error Handler");
            Console.WriteLine(new string('=', 60));

            async Task<ChatCompletion?> SafeApiCall(string model, string userMessage, int maxTokens = 10000)
            {
                var client = new OpenAIClient(new ApiKeyCredential(ApiKey), new OpenAIClientOptions
                {
                    Endpoint = new Uri(BaseUrl)
                });

                var chatClient = client.GetChatClient(model);

                try
                {
                    var response = await chatClient.CompleteChatAsync(
                        messages: new[] { new UserChatMessage(userMessage) },
                        options: new ChatCompletionOptions { MaxOutputTokenCount = maxTokens }
                    );
                    return response.Value;
                }
                catch (Exception ex)
                {
                    string errorMsg = ex.Message.ToLower();

                    if (errorMsg.Contains("authentication") || errorMsg.Contains("api key"))
                    {
                        Console.WriteLine($"✗ Authentication Error: {ex.Message}");
                        Console.WriteLine("Check your API key.");
                    }
                    else if (errorMsg.Contains("not found") || errorMsg.Contains("model"))
                    {
                        Console.WriteLine($"✗ Not Found: {ex.Message}");
                        Console.WriteLine("Model may not exist or be unavailable.");
                    }
                    else if (errorMsg.Contains("rate limit"))
                    {
                        Console.WriteLine($"⚠ Rate Limit: {ex.Message}");
                        Console.WriteLine("Too many requests. Please wait and try again.");
                    }
                    else if (errorMsg.Contains("timeout"))
                    {
                        Console.WriteLine($"⚠ Timeout: {ex.Message}");
                        Console.WriteLine("Request took too long. Try again.");
                    }
                    else if (errorMsg.Contains("bad request"))
                    {
                        Console.WriteLine($"✗ Bad Request: {ex.Message}");
                        Console.WriteLine("Check your request parameters.");
                    }
                    else
                    {
                        Console.WriteLine($"✗ Error: {ex.Message}");
                    }

                    return null;
                }
            }

            // Test the wrapper
            Console.WriteLine("\nTesting with valid request:");
            var result = await SafeApiCall("gpt-5-nano", "Say hello in 3 words");

            if (result != null)
            {
                Console.WriteLine($"✓ Success: {result.Content[0].Text}");
            }
            else
            {
                Console.WriteLine("Request failed gracefully");
            }
        }

        /// <summary>
        /// Validate inputs before making API request
        /// </summary>
        private static async Task ValidateBeforeRequest()
        {
            Console.WriteLine("\n" + new string('=', 60));
            Console.WriteLine("Input Validation");
            Console.WriteLine(new string('=', 60));

            async Task<ChatCompletion?> ValidateAndCall(string model, string userInput)
            {
                // Validate model name
                var validModels = new[] { "gpt-5-nano", "claude-haiku-4-5", "deepseek-chat", "gemini-2.5-flash-lite" };

                if (!validModels.Contains(model))
                {
                    Console.WriteLine($"✗ Invalid model: {model}");
                    Console.WriteLine($"Valid models: {string.Join(", ", validModels)}");
                    return null;
                }

                // Validate user input
                if (string.IsNullOrWhiteSpace(userInput))
                {
                    Console.WriteLine("✗ Empty user input");
                    return null;
                }

                if (userInput.Length > 10000)
                {
                    Console.WriteLine("✗ Input too long (max 10,000 characters)");
                    return null;
                }

                // All validations passed
                Console.WriteLine("✓ Validation passed");

                try
                {
                    var client = new OpenAIClient(new ApiKeyCredential(ApiKey), new OpenAIClientOptions
                    {
                        Endpoint = new Uri(BaseUrl)
                    });

                    var chatClient = client.GetChatClient(model);

                    var response = await chatClient.CompleteChatAsync(
                        messages: new[] { new UserChatMessage(userInput) },
                        options: new ChatCompletionOptions { MaxOutputTokenCount = 10000 }
                    );
                    return response.Value;
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"✗ Error: {ex.Message}");
                    return null;
                }
            }

            // Test validation
            Console.WriteLine("\nTest 1: Invalid model");
            await ValidateAndCall("invalid-model", "Hello");

            Console.WriteLine("\nTest 2: Empty input");
            await ValidateAndCall("gpt-5-nano", "");

            Console.WriteLine("\nTest 3: Valid request");
            var result = await ValidateAndCall("gpt-5-nano", "Say hi");
            if (result != null)
            {
                Console.WriteLine($"Response: {result.Content[0].Text}");
            }
        }
    }
}
